utils/apiAuth.js

```diff
@@ -0,0 +1,91 @@
+'use strict';
+
+const AuthenticationError = require('./controllerUtils').AuthenticationError;
+const { handleErrors } = require('./controllerUtils');
+const { db, pgp } = require('./dbUtils');
+const logger = require('./logger').logger;
+
+const qrec = pgp.errors.queryResultErrorCode;
+
+const tokenColsCreate = new pgp.helpers.ColumnSet(['auth_token', 'username'], {
+  table: 'token',
+});
+
+const addNewKey = async (req, res, next) => {
+  try {
+    //authenticate api-key from header before continuing
+    let apiKey = req.header('X-API-KEY');
+    const keyExists = await authenticateKey(apiKey);
+
+    console.log('key in header good');
+
+    const newKey = await genAPIKey(req.body.username);
+
+    const returnKey = {
+      'X-API-KEY': newKey.authToken,
+    };
+    return res.status(201).json(returnKey);
+  } catch (e) {
+    handleErrors(e, 'Error adding sign-off', logger, next);
+  }
+};
+
+const genAPIKey = (user) => {
+  //create a base-36 string that contains 30 chars in a-z,0-9
+  const APItoken = [...Array(30)]
+    .map((e) => ((Math.random() * 36) | 0).toString(36))
+    .join('');
+
+  console.log(APItoken);
+
+  const tokenData = {
+    auth_token: APItoken,
```

Comment:
Token should be encrypted/hashed for storage. Reversible encryption means we can recover the API keys, but we need to also have a key for decrypting the API keys. Hashing means we can't recover the API keys, but that doesn't really matter since we can just generate a new one. Hashing is simpler and I don't think encrypting has any benefits here